### PR TITLE
CIRC-1588 Fix NPE, add tests

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose 
- Fix NPE in Items In Transit report;
- Add tests;

Resolves: [CIRC-1588](https://issues.folio.org/browse/CIRC-1588)

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java [hunk 1]
```diff
@@ -68,15 +68,19 @@ private JsonObject buildEntry(Item item) {
     Location location = reportContext.getLocations().get(item.getEffectiveLocationId());
     if (location != null) {
       ServicePoint primaryServicePoint = reportContext.getServicePoints()
-        .get(location.getPrimaryServicePointId().toString());
+        .get(location.getPrimaryServicePointId() != null
+          ? location.getPrimaryServicePointId().toString()
+          : null);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java:73]:
```
    Optional.ofNullable(item.getLastCheckInServicePointId())
      .map(UUID::toString)
      .map(reportContext.getServicePoints()::get)
      .ifPresent(item::updateLastCheckInServicePoint);
```

File: src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java [hunk 2]
```diff
@@ -68,15 +68,19 @@ private JsonObject buildEntry(Item item) {
...
         .withLocation(location.withPrimaryServicePoint(primaryServicePoint));
     }
 
     ServicePoint inTransitDestinationServicePoint = reportContext.getServicePoints()
       .get(item.getInTransitDestinationServicePointId());
     ServicePoint lastCheckInServicePoint = reportContext.getServicePoints()
-      .get(item.getLastCheckInServicePointId().toString());
+      .get(item.getLastCheckInServicePointId() != null
+        ? item.getLastCheckInServicePointId().toString()
+        : null);
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java:83]: Try use Optional

File: src/test/java/api/requests/ItemsInTransitReportTests.java [hunk 3]
```diff
@@ -469,6 +470,50 @@ void reportWillNotFailWithUriTooLargeError() {
     assertThat(itemsInTransitReport.size(), is(200));
   }
 
+  @Test
+  void reportShouldNotFailWithoutLastServicePointId() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/ItemsInTransitReportTests.java:474]: Add test for missing `lastCheckIn`

File: src/test/java/api/requests/ItemsInTransitReportTests.java [hunk 4]
```diff
@@ -469,6 +470,50 @@ void reportWillNotFailWithUriTooLargeError() {
...
+    final UUID firstServicePointId = servicePointsFixture.cd1().getId();
+    final UUID forthServicePointLocationId = locationsFixture.fourthServicePoint().getId();
+
+    ItemResource item = createSmallAngryPlanetCopy(forthServicePointLocationId, "111");
+
+    checkOutFixture.checkOutByBarcode(item);
+    checkInFixture.checkInByBarcode(new CheckInByBarcodeRequestBuilder()
+      .forItem(item)
+      .on(ClockUtil.getZonedDateTime())
+      .at(firstServicePointId));
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/ItemsInTransitReportTests.java:513]: assert that item is in transit

File: src/test/java/api/requests/ItemsInTransitReportTests.java [hunk 5]
```diff
@@ -469,6 +470,50 @@ void reportWillNotFailWithUriTooLargeError() {
...
+    List<JsonObject> itemsInTransitReport = ResourceClient.forItemsInTransitReport().getAll();
+
+    assertThat(itemsInTransitReport.size(), is(1));
+  }
+
+  private ItemResource checkOutAndCheckInItem() {
+    final UUID firstServicePointId = servicePointsFixture.cd1().getId();
+    final UUID forthServicePointLocationId = locationsFixture.fourthServicePoint().getId();
+
+    ItemResource item = createSmallAngryPlanetCopy(forthServicePointLocationId, "111");
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/ItemsInTransitReportTests.java:507]: create item without parameters, check in at service point != cd1

File: src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java [hunk 6]
```diff
@@ -100,6 +105,7 @@ private JsonObject buildEntry(Item item) {
     write(entry, "effectiveCallNumberComponents",
       createCallNumberComponents(item.getCallNumberComponents()));
 
+    ServicePoint inTransitDestinationServicePoint = item.getInTransitDestinationServicePoint();
```
Comment (OleksandrVidinieiev) [inline src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java:108]: What changed here?

File: src/test/java/api/requests/ItemsInTransitReportTests.java [hunk 7]
```diff
@@ -469,6 +470,63 @@ void reportWillNotFailWithUriTooLargeError() {
     assertThat(itemsInTransitReport.size(), is(200));
   }
 
+  @Test
+  void reportShouldNotFailWithoutLastServicePointId() {
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/ItemsInTransitReportTests.java:474]:
```suggestion
  void reportShouldNotFailWithoutLastCheckInServicePointId() {
```

File: src/test/java/api/requests/ItemsInTransitReportTests.java [hunk 8]
```diff
@@ -469,6 +470,63 @@ void reportWillNotFailWithUriTooLargeError() {
...
+
+  @Test
+  void reportShouldNotFailWithoutPrimaryServicePointId() {
+    ItemResource item = checkOutAndCheckInItem();
+
+    Response response = itemsClient.getById(item.getId());
+    JsonObject checkedInItemJson = response.getJson();
+    UUID permanentLocationId = UUID.fromString(checkedInItemJson.getString("permanentLocationId"));
+    JsonObject location = locationsClient.getById(permanentLocationId).getJson();
+    location.put("primaryServicePoint", null);
```
Comment (OleksandrVidinieiev) [inline src/test/java/api/requests/ItemsInTransitReportTests.java:495]:
```suggestion
    location.putNull("primaryServicePoint")
```

File: src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java [hunk 9]
```diff
@@ -67,20 +68,24 @@ private JsonObject buildEntry(Item item) {
...
     }
 
-    ServicePoint inTransitDestinationServicePoint = reportContext.getServicePoints()
-      .get(item.getInTransitDestinationServicePointId());
-    ServicePoint lastCheckInServicePoint = reportContext.getServicePoints()
-      .get(item.getLastCheckInServicePointId().toString());
+    item = ofNullable(item.getInTransitDestinationServicePointId())
+      .map(reportContext.getServicePoints()::get)
+      .map(item::updateDestinationServicePoint)
+      .orElse(item);
```
[the remark this answers is not in the document]
Comment (roman-barannyk) [reply src/main/java/org/folio/circulation/domain/representations/ItemsInTransitReport.java:82]: No, it returns a new instance

File: src/test/java/api/requests/ItemsInTransitReportTests.java (hunk 7 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/ItemsInTransitReportTests.java]: renamed

File: src/test/java/api/requests/ItemsInTransitReportTests.java (hunk 8 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/ItemsInTransitReportTests.java]: updated

File: src/test/java/api/requests/ItemsInTransitReportTests.java (hunk 5 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/ItemsInTransitReportTests.java:522]: it doesn't work without parameters, we should specify a service point

File: src/test/java/api/requests/ItemsInTransitReportTests.java (hunk 4 above)
Comment (roman-barannyk) [reply src/test/java/api/requests/ItemsInTransitReportTests.java:527]: done

**Review by OleksandrVidinieiev — APPROVED** [review]

**Pull request merged**